Libraries/LibWeb/CMakeLists.txt

```diff
@@ -142,21 +142,31 @@ function(libweb_js_wrapper class)
     get_filename_component(basename ${class} NAME)
     add_wrapper_sources(Bindings/${basename}Wrapper.cpp Bindings/${basename}Wrapper.h)
     add_custom_command(
-        OUTPUT Bindings/${basename}Wrapper.h
+        OUTPUT Bindings/${basename}Wrapper.h.tmp
         COMMAND /bin/mkdir -p Bindings
-        COMMAND WrapperGenerator --header ${CMAKE_CURRENT_SOURCE_DIR}/${class}.idl > Bindings/${basename}Wrapper.h
+        COMMAND WrapperGenerator --header ${CMAKE_CURRENT_SOURCE_DIR}/${class}.idl > Bindings/${basename}Wrapper.h.tmp
         VERBATIM
         DEPENDS WrapperGenerator
         MAIN_DEPENDENCY ${class}.idl
     )
     add_custom_command(
-        OUTPUT Bindings/${basename}Wrapper.cpp
+        OUTPUT Bindings/${basename}Wrapper.h
+        COMMAND ${CMAKE_COMMAND} -E copy_if_different Bindings/${basename}Wrapper.h.tmp Bindings/${basename}Wrapper.h
+        DEPENDS Bindings/${basename}Wrapper.h.tmp
+    )
+    add_custom_command(
+        OUTPUT Bindings/${basename}Wrapper.cpp.tmp
         COMMAND /bin/mkdir -p Bindings
-        COMMAND WrapperGenerator --implementation ${CMAKE_CURRENT_SOURCE_DIR}/${class}.idl > Bindings/${basename}Wrapper.cpp
+        COMMAND WrapperGenerator --implementation ${CMAKE_CURRENT_SOURCE_DIR}/${class}.idl > Bindings/${basename}Wrapper.cpp.tmp
         VERBATIM
         DEPENDS WrapperGenerator
         MAIN_DEPENDENCY ${class}.idl
     )
+    add_custom_command(
+        OUTPUT Bindings/${basename}Wrapper.cpp
+        COMMAND ${CMAKE_COMMAND} -E copy_if_different Bindings/${basename}Wrapper.cpp.tmp Bindings/${basename}Wrapper.cpp
+        DEPENDS Bindings/${basename}Wrapper.cpp.tmp
+    )
```

Comment:
It seems like this logic could/should be encapsulated into a single cmake function, like serenity_bin/serenity_lib

Reply:
I don't know, the plan is to not have this in cmake code eventually anyways. And it's not immediately clear to me how to do custom commands with optionally more than one COMMAND from a function. So since it's (for me) not easy to do and temporary, I deemed it not worth it.

Reply:
> the plan is to not have this in cmake code eventually anyways.

Care to elaborate?

Reply:
See "It'd be nicer if the code generators did this "write
output only if it's different from what's on disk" optimization
themselves, generating the potentially-written output only in
memory. " above.

Reply:
*poke* See #2970. Does that solve the problem as intended? :)

Reply:
I think it's nicer if the generator generates stuff in memory and only writes to disk if it would write something new, but I also think it's even more nicer if someone else does it, so feel free to take over ;)

Reply:
> in memory

Yes, but then the comparison would also need to happen in-memory, and that makes things unnecessarily complicated. At least I don't know how to do that in `sh` :P

So all you've gotta do now is wait until #2970 lands.

Reply:
Well, it needs cooperation from each generator (http://reviews.llvm.org/rL373664 is a real-world example). #2970 means the approach in this change here doesn't permanently leave the .tmp file on disk which is nice, but you still need an extra build edge, you still have the disk traffic, and you still need the cmake goop. I'm always happy when I can move logc out of the build description into an actual programming language :) (But #2970 still definitely makes this change here nicer.)

Reply:
Hmm? #2970 does not need further cooperation from each generator. (They all already write to stdout.) I agree that not touching the disk would be even nicer, I'll spend some more minutes trying to get that to work (but not hours, because this affects only a few files).

@cmake goop: Hmm? With the script, you only need a few more characters: `path/to/Meta/write-only-if-changed.sh `.

Reply:
Having to author generators to not save output if it's unchanged seems like increasing the complexity of writing a generator for quite little benefit, no?

Reply:
Ah, I misread how #2970 is used. That makes everything much nicer indeed.